File: backend/src/modules/order/application/use-cases/get-all-orders.usecase.ts
import { Inject } from '@nestjs/common';
import { ORDER_REPOSITORY } from '../../domain/repositories/order.repository';
import type { OrderRepository } from '../../domain/repositories/order.repository';
import { UseCase } from '@shared/application/usecase.interface';
import type { OutputPort } from '@shared/application/contracts/output-port.interface';
import { GetAllOrdersRequest } from '../contracts/requests/get-all-orders.request';
import { GetAllOrdersResponse } from '../contracts/responses/get-all-orders.response';

export class GetAllOrdersUseCase implements UseCase<GetAllOrdersRequest> {
  constructor(
    private readonly outputPort: OutputPort<GetAllOrdersResponse>,
    @Inject(ORDER_REPOSITORY) private readonly orders: OrderRepository,
  ) {}

  async execute(_: GetAllOrdersRequest): Promise<void> {
    const orders = await this.orders.getAllOrders();
    this.outputPort.present({ orders });
  }
}

File: frontend/lib/features/order/domain/usecases/get_all_orders_usecase.dart
import '../../../../core/domain/usecase.dart';
import '../../../../core/utils/result.dart';
import '../../domain/entities.dart';
import '../../domain/repositories/order_repository.dart';

/// Use case: Get All Orders
class GetAllOrdersUseCase implements UseCaseNoInput<Result<List<Order>>> {
  final OrderRepository _repository;

  GetAllOrdersUseCase(this._repository);

  @override
  Future<Result<List<Order>>> call() {
    return _repository.getAllOrders();
  }
}
